binaries/daemon/src/lib.rs

```diff
@@ -1771,6 +1849,96 @@ impl Daemon {
+            format!("propagate error failed: no running dataflow with ID `{dataflow_id}`")
+        })?;
+
+        tracing::warn!(
+            "Node {}/{} failed: {}. Propagating error to downstream nodes.",
+            dataflow_id,
+            source_node_id,
+            error_message
+        );
+
+        // Get all outputs of the failed node
+        let outputs: Vec<DataId> = dataflow
+            .mappings
+            .keys()
+            .filter(|m| &m.0 == &source_node_id)
+            .map(|m| m.1.clone())
+            .collect();
+
+        // Group affected inputs by receiver node
+        let mut affected_by_receiver: BTreeMap<NodeId, Vec<DataId>> = BTreeMap::new();
+
+        for output_id in &outputs {
+            let output_key = OutputId(source_node_id.clone(), output_id.clone());
+            if let Some(receivers) = dataflow.mappings.get(&output_key) {
+                for (receiver_node_id, input_id) in receivers {
+                    affected_by_receiver
+                        .entry(receiver_node_id.clone())
+                        .or_insert_with(Vec::new)
+                        .push(input_id.clone());
+                }
+            }
+        }
+
+        // Check if there are any remote receivers (nodes not on this daemon)
+        let has_remote_receivers = affected_by_receiver
+            .keys()
+            .any(|receiver_node_id| !dataflow.subscribe_channels.contains_key(receiver_node_id));
+
+        // Send NodeFailed event to each local receiver
+        for (receiver_node_id, affected_input_ids) in affected_by_receiver {
+            if let Some(channel) = dataflow.subscribe_channels.get(&receiver_node_id) {
+                // Local receiver - send directly
+                let event = NodeEvent::NodeFailed {
+                    affected_input_ids,
+                    error: error_message.clone(),
+                    source_node_id: source_node_id.clone(),
+                };
+                let _ = send_with_timestamp(channel, event, &self.clock);
+            }
+        }
+
+        // If there are remote receivers, send a single InterDaemonEvent
+        // The receiving daemon will determine which of its nodes are affected
+        if has_remote_receivers {
+            // Use the first output as the routing key (any output will work)
+            let first_output = outputs.first().ok_or_else(|| {
+                eyre!("Failed node has no outputs, cannot send error to remote receivers")
+            })?;
+            let output_id = OutputId(source_node_id.clone(), first_output.clone());
+            let event = InterDaemonEvent::NodeFailed {
+                dataflow_id,
+                source_node_id: source_node_id.clone(),
+                error: error_message.clone(),
+            };
+
+            // Send to remote receivers using the same mechanism as regular outputs
+            if let Err(err) = self
+                .send_to_remote_receivers(dataflow_id, &output_id, event)
```

Comment:
I forgot that daemons subscribe to a specific output topic. So just sending the message to the first output is not enough unfortunately (as some daemon might only be subscribed to the second or third output).

So I guess we have two options:

- send the message to all outputs and deduplicate it on the receiver side (i.e. remember which nodes failed already to avoid sending duplicated events to nodes)
- add a new node-level subscription topic for such messages that all affected daemons subscribe to

What do you prefer?